examples/control-vector-generator/control-vector-generator.cpp

```diff
@@ -0,0 +1,881 @@
+
+    // no need to load anything into square or square_transpose yet
+
+    // initialize model.eigenvector to random vector
+    std::vector<float> random_vec;
+    std::default_random_engine generator(static_cast<unsigned int>(std::time(0)));
+    std::uniform_real_distribution<float> distribution(0.0, 1.0);
+    for (int i = 0; i < v_diff_original->ne[1]; ++i) {
+        random_vec.push_back(distribution(generator));
+    }
+
+    // we don't normalize it at first but that shouldn't be a problem
+    ggml_backend_tensor_set(model.eigenvector, random_vec.data(), 0, ggml_nbytes(model.eigenvector));
+}
+
+struct ggml_cgraph * square_diff_graph(const pca_model & model) {
+    static size_t buf_size = ggml_tensor_overhead() * GGML_DEFAULT_GRAPH_SIZE + ggml_graph_overhead();
+    static std::vector<uint8_t> buf(buf_size);
+
+    struct ggml_init_params params0 = {
+        /*.mem_size   =*/ buf_size,
+        /*.mem_buffer =*/ buf.data(),
+        /*.no_alloc   =*/ true, // the tensors will be allocated later by ggml_allocr_alloc_graph()
+    };
+    struct ggml_context * ctx0 = ggml_init(params0);
+    struct ggml_cgraph  * gf = ggml_new_graph(ctx0);
+
+    struct ggml_tensor * square = ggml_mul_mat(ctx0, model.v_diff_original, model.v_diff_original);
+    //struct ggml_tensor * square_transpose = ggml_transpose(ctx0, square);
+
+    ggml_build_forward_expand(gf, square);
+
+    ggml_free(ctx0);
+    return gf;
+}
+
+struct ggml_tensor * compute_square(const pca_model & model, ggml_gallocr_t allocr, int n_threads) {
+    struct ggml_cgraph * gf = square_diff_graph(model);
+
+    ggml_gallocr_alloc_graph(allocr, gf);
+
+    if (ggml_backend_is_cpu(model.backend)) {
+        ggml_backend_cpu_set_n_threads(model.backend, n_threads);
+    }
+
+#ifdef GGML_USE_METAL
+    if (ggml_backend_is_metal(model.backend)) {
+        ggml_backend_metal_set_n_cb(model.backend, n_threads);
+    }
+#endif
+
+    ggml_backend_graph_compute(model.backend, gf);
+
+    return gf->nodes[gf->n_nodes - 1];
+}
+
+struct ggml_cgraph * power_iteration_graph(const pca_model & model, float tolerance) {
+    static size_t buf_size = ggml_tensor_overhead() * GGML_DEFAULT_GRAPH_SIZE + ggml_graph_overhead();
+    static std::vector<uint8_t> buf(buf_size);
+
+    struct ggml_init_params params0 = {
+        /*.mem_size   =*/ buf_size,
+        /*.mem_buffer =*/ buf.data(),
+        /*.no_alloc   =*/ true, // the tensors will be allocated later by ggml_allocr_alloc_graph()
+    };
+    struct ggml_context * ctx0 = ggml_init(params0);
+    struct ggml_cgraph  * gf = ggml_new_graph(ctx0);
+
+    struct ggml_tensor * b_tensor = ggml_mul_mat(ctx0, model.square, model.eigenvector);
+    // TODO difference between ggml_norm and ggml_norm_inplace?
+    // also is this the right way to do multi-step graphs?
+    b_tensor = ggml_norm_inplace(ctx0, b_tensor, tolerance);
+
+    ggml_build_forward_expand(gf, b_tensor);
+
+    ggml_free(ctx0);
```

Comment:
hmm be careful here, the `ggml_cgraph` is created inside ctx0, but we free ctx0 then return gf. That will makes gf to be freed along with ctx0.

Can we create a `pca_ctx` in `main()`, then pass it into this function?

Reply:
Yes, that shouldn't be too difficult. Additionally we should know exactly how much memory to allocate because we know the dimensions of the matrices and vectors ahead of time, and how many layers to work with.